src/core/simulation.jl

```diff
@@ -225,23 +242,47 @@ function _check_steps(
     stage_initial_times::Dict{Int, Vector{Dates.DateTime}},
 )
     for (stage_number, stage_name) in sim.sequence.order
-        forecast_count = length(stage_initial_times[stage_number])
         stage = sim.stages[stage_name]
-        if get_steps(sim) * get_executions(stage) > forecast_count
+        execution_counts = get_executions(stage)
+        @assert length(findall(x -> x == stage_number, sim.sequence.execution_order)) ==
+                execution_counts
+        forecast_count = length(stage_initial_times[stage_number])
+        if get_steps(sim) * execution_counts > forecast_count
             throw(IS.ConflictingInputsError("The number of available time series ($(forecast_count)) is not enough to perform the
             desired amount of simulation steps ($(sim.steps*stage.internal.execution_count))."))
         end
     end
     return
 end
 
+function _check_required_ini_cond_caches(sim::Simulation)
+    for (stage_number, stage_name) in sim.sequence.order
+        stage = get_stage(sim, stage_name)
+        for (k, v) in get_initial_conditions(stage.internal.psi_container)
+            # No cache needed for the initial condition -> continue
+            isnothing(v[1].cache_type) && continue
+            c = get_cache(sim, v[1].cache_type, k.device_type)
+            if isnothing(c)
+                throw(ArgumentError("Cache $(v[1].cache_type) not defined for initial condition $(k) in stage $stage_name"))
+            end
+            @debug "found cache $(v[1].cache_type) for initial condition $(k) in stage $(stage_name)"
+        end
+    end
+    return
+end
+
 function _populate_caches!(sim::Simulation, stage_name::String)
-    caches = get(sim.sequence.cache, stage_name, nothing)
+    caches = get_stage_cache_definition(sim, stage_name)
     isnothing(caches) && return
-    cache_dict = Dict{Type{<:AbstractCache}, AbstractCache}()
+    stage = get_stage(sim, stage_name)
     for c in caches
-        sim.stages[stage_name].internal.cache_dict[typeof(c)] = c
-        build_cache!(c, sim.stages[stage_name].internal.psi_container)
+        cache_key = CacheKey(c)
+        push!(stage.internal.caches, cache_key)
+        if !haskey(sim.internal.simulation_cache, cache_key)
```

Comment:
This code block should probably be a function in SimulationInternal.
`function update_cache_value(internal::SimulationInternal, c, value)`

That would keep the implementation details of SimulationInternal within itself.